### PR TITLE
add subtotal and payment date to invoices

### PR DESCRIPTION
**Pull request by cayacdev**
## Description
A clear and concise description of what this pull request adds or changes.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist
- [x] I verified that my code applies to the guidelines (`composer code-check`)
- [ ] I updated the documentation (see [here](https://github.com/kimai/www.kimai.org/tree/master/_documentation))
- [x] I agree that this code is used in Kimai and will be published under the [MIT license](https://github.com/kevinpapst/kimai2/blob/master/LICENSE)

## Open Todos
- [x] Adapt entity and setup migrations
- [x] Create setPaymentDate form
- [x] Add new column to invoice listing: subtotal
- [x] Add new column to invoice listing: payment date
- [x] Add subtotal to export
- [x] Add payment date to export
- [x] Translations
- [x] Unit Tests

### CODE REVIEW

File: src/EventSubscriber/Actions/InvoiceSubscriber.php
Comment (kevinpapst) [inline src/EventSubscriber/Actions/InvoiceSubscriber.php:34]: What is your thought on this change?

Comment (kevinpapst) [inline src/EventSubscriber/Actions/InvoiceSubscriber.php:34]: What is your thought on this change?

**Review by kevinpapst — CHANGES_REQUESTED** [review]
Thanks, looks very good 👍 

Here is my first round of review comments, I haven't tested it yet because I am working on other invoice changes as well and can't commit right away. But I'll do that a little bit later today

**Comment by kevinpapst** [comment]
Oh and please run `composer codestyle-fix` before committing, so the CI can test the next stage

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php:58]: I'm not sure if I need the option `time_increment` anyway.

Comment (kevinpapst) [reply src/Form/InvoiceSetPaymentDateForm.php:58]: skip it, totally fine to set it to 30 fixed

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: renamed

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: When I remove this option then I get `An error has occurred resolving the options of the form "App\Form\InvoicePaymentDateForm": The option "time_increment" does not exist.`

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: done

File: src/Invoice/ServiceInvoice.php
Comment (cayacdev) [reply src/Invoice/ServiceInvoice.php]: done

File: templates/invoice/listing.html.twig
Comment (cayacdev) [reply templates/invoice/listing.html.twig]: done

File: templates/invoice/set_payment_date.html.twig
Comment (cayacdev) [reply templates/invoice/set_payment_date.html.twig]: done

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: done

File: src/Controller/InvoiceController.php
Comment (cayacdev) [reply src/Controller/InvoiceController.php]: done

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: done

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: done

File: templates/invoice/payment_date_edit.html.twig
Comment (cayacdev) [inline templates/invoice/payment_date_edit.html.twig:10]: - [ ] Note for myself: Change back path

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (kevinpapst) [reply src/Form/InvoiceSetPaymentDateForm.php]: Thats because you declare it in `createPaymentDateForm()` in the controller. 
Remove it at both places and then register it directly with the DateTimePickerType:
```
            ->add('paymentDate', DateTimePickerType::class, array_merge($dateTimeOptions, [
                 'label' => 'label.paymentDate',
                 'required' => true,
                 'time_increment' => 30,
             ]));
```

File: src/Migrations/Version20210320162820.php
Comment (cayacdev) [reply src/Migrations/Version20210320162820.php:33]: good question. I haven't even thought about that yet. I test this later, I guess also that mysql set existing columns to `0`. Maybe it is better to update existing columns?

- [ ] Note for myself: Test what happen with existing rows when applying the new migration

Comment (kevinpapst) [reply src/Migrations/Version20210320162820.php:33]: Probably add another migration to set `subtotal = total - tax where subtotal = 0` ... but actually now that I wrote this, why do we need a new column? Can't this be a calculated value in the Invoice entity?

File: src/EventSubscriber/Actions/InvoiceSubscriber.php
Comment (cayacdev) [reply src/EventSubscriber/Actions/InvoiceSubscriber.php:34]: In testing, I created a lot of invoices to go through the states. It would be much easier if you could set invoices in the status `paid` back to `pending.` 

Then I thought this is a valid use case also for prod environments. For example, if you accidentally chose a wrong payment date, or set the wrong invoice to `paid` and you want to correct that mistake.

File: templates/invoice/set_payment_date.html.twig
Comment (cayacdev) [reply templates/invoice/set_payment_date.html.twig]: When I apply this change, then the save button of the form wont submit it.

File: src/Controller/InvoiceController.php
Comment (cayacdev) [reply src/Controller/InvoiceController.php]: done

File: src/Migrations/Version20210320162820.php
Comment (cayacdev) [reply src/Migrations/Version20210320162820.php:33]: Yes, I think that makes sense. This means that I have only a getter (no setter) in the `App/Entity/Invoice.php` with the following content:

```
    public function getSubtotal(): ?float
    {
        return $this->total - $this->tax;
    }
```
Also I would remove the column in the migration file?
Correct?

Comment (kevinpapst) [reply src/Migrations/Version20210320162820.php:33]: I am still thinking... so in theory and for your use-case this should be sufficient, correct? 
And we can use SQL if we need would need this value in queries (which we don't need yet).
There are more ideas for the future (like multiple tax rates per invoice, discounts and such) which might influence this value, but I'd say KISS here and don't built for an unknown future.

Comment (cayacdev) [reply src/Migrations/Version20210320162820.php]: done

File: templates/invoice/macros.html.twig
Comment (cayacdev) [reply templates/invoice/macros.html.twig]: done

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: done, but can you explain to me what `time_increment` does?

Comment (kevinpapst) [reply src/Form/InvoiceSetPaymentDateForm.php]: It defines the minutes in the picker, where you choose the minutes via dropdown. Normally you can choose from 0-59 every single minute, which I believe is not necessary here. You can also delete the option completely, I guess most people will not choose an exact minute there. What do you think?

File: src/Entity/Invoice.php
Comment (kevinpapst) [inline src/Entity/Invoice.php:178]: I guess you want to add a new translation, correct? In that case we need to change the label later (and remove the second *)

File: src/Form/InvoiceSetPaymentDateForm.php
Comment (cayacdev) [reply src/Form/InvoiceSetPaymentDateForm.php]: Ok, I guess I can remove it. I modeled the payment_date as `type: 'date' , so it has no time information. Maybe the types  in the entity class are wrong. But I don't know which type can be used instead of `DateTime`. `Date` seems not to exist.

File: src/Entity/Invoice.php
Comment (cayacdev) [reply src/Entity/Invoice.php]: done

File: templates/invoice/payment_date_edit.html.twig
Comment (cayacdev) [reply templates/invoice/payment_date_edit.html.twig]: done

**Comment by cayacdev** [comment]
Hi @kevinpapst,

I think I'm almost done. The last thing is to create the unit tests for `InvoiceSubscriber.php`, `InvoicePaymentDateForm.php` and to adapt the documentation I guess.
But I don't know where to document this new feature. On tis page https://www.kimai.org/documentation/invoices.html the invoice states (new, waiting for payment, paid) are not documented as well.

Also I need little help with the unit test for the InvoiceController. I can't manage to create a valid request with valid form data (tests/Controller/InvoiceControllerTest.php line 357).

**Comment by kevinpapst** [comment]
I will give you feedback, as soon as possible. 
But I doubt that this will happen before the weekend, because I am terrible busy currently.
Sorry, but we will get there soon!

**Comment by cayacdev** [comment]
> I will give you feedback, as soon as possible.
> But I doubt that this will happen before the weekend, because I am terrible busy currently.
> Sorry, but we will get there soon!

No hurry ;)

**Review by kevinpapst — DISMISSED** [review]
I applied some last changes after testing, if you want to double-check: now is the time :D

**Comment by cayacdev** [comment]
Cool, thanks a lot! Should I add the payment date into the search? What do you think?

Screenshot 2021-03-28 at 17 58 58

**Comment by cayacdev** [comment]
Ok, I tried it, but I didn't find the place where the date range is applied to a specific invoice attribute (createdAt I guess). Maybe I will provide a PR for this later. 

So I change the state of this PR to ready? 

In the next weeks I will create the PR for the custom field support for invoices ;)

**Comment by kevinpapst** [comment]
Almost every repository in Kimai has such a method where the doctrine / SQL query is built:
https://github.com/kevinpapst/kimai2/blob/master/src/Repository/InvoiceRepository.php#L154

Should be rather simple, add an optional [daterange](https://github.com/kevinpapst/kimai2/blob/master/src/Form/Toolbar/AbstractToolbarForm.php#L149) to the [form](https://github.com/kevinpapst/kimai2/blob/master/src/Form/Toolbar/InvoiceArchiveForm.php) (as you already did), add a [use DateRangeTrait](https://github.com/kevinpapst/kimai2/blob/master/src/Repository/Query/DateRangeTrait.php) to the  [InvoiceArchiveQuery](https://github.com/kevinpapst/kimai2/blob/master/src/Repository/Query/InvoiceArchiveQuery.php) and then the last piece is to add begin and end date as `SQL where` parts in the [repository method](https://github.com/kevinpapst/kimai2/blob/master/src/Repository/InvoiceRepository.php#L154).

**Comment by cayacdev** [comment]
I had to force push because my commits weren't signed ;)

File: src/Entity/Invoice.php
Comment (kevinpapst) [inline src/Entity/Invoice.php:33]: Can you try if this works and removes the `00:00:00` from the export file?
```suggestion
 * @Exporter\Expose("paymentDate", label="invoice.payment_date", type="date", exp="object.getPaymentDate() === null ? null : object.getPaymentDate()")
```

Comment (cayacdev) [reply src/Entity/Invoice.php]: works -> commited it.

**Review by kevinpapst — APPROVED** [review]
Thanks @cayacdev 🎉 for the PR!

Looking forward for more 😁

**Pull request merged**
